lib/github.com/diku-dk/sparse/blocked_square_regular.fut

```diff
@@ -14,322 +14,345 @@ local
+         in y
 
   module lu = mk_lu (T)
 
   def matsub a b = linalg.matop (T.-) a b
 
-  def blu_nofill [n] (a:mat[n]) : mat[n] =
-    let [nz] a : {idxs:[nz]i64,blks:[nz][bsz][bsz]t,n:[n]()} = a
+  def blu_nofill [n] (a: mat [n]) : mat [n] =
+    let [nz] a: {idxs: [nz]i64, blks: [nz][bsz][bsz]t, n: [n]()} = a
     let idxs = a.idxs
     let blks = a.blks
-    let nb = n / bsz   -- number of blocks in each dimension
-    let hrcbs = map3 (\h i b -> let (r,c) = idx_unflatten nb i
-                                in (h,r,c,b))
-                     (iota nz) idxs blks
+    let nb = n / bsz
+    -- number of blocks in each dimension
+    let hrcbs =
+      map3 (\h i b ->
+              let (r, c) = idx_unflatten nb i
+              in (h, r, c, b))
+           (iota nz)
+           idxs
+           blks
     let hrcbs =
-      loop hrcbs for i < nb do
-      let blks = filter (\(_,r,c,_) -> r >= i && c >= i) (copy hrcbs)
-      let (h,r,c,b) = blks[0] -- h is the idx into hrcbs identifying the current diagonal block
-      let b = lu.lu 1 b
-      in assert (r == i && c == i)
-                (let A21 = filter (\(_,r,c,_) -> r > i && c == i) blks
-                 let A12 = filter (\(_,r,c,_) -> c > i && r == i) blks
-                 let X21 = map (\(h,r,_,a) -> (h,r,map (\aa -> backsolve' aa b) a)) A21
-                 let X12 = map (\(h,_,c,a) -> (h,c,transpose(map (\aa -> forsolve b aa) (transpose a)))) A12
-                 let hrcbs[h] = (h,r,c,b)
-                 let hrcbs =
-                   loop hrcbs = hrcbs for j < length X21 do
-                   let (h,r,b) = X21[j]
-                   let hrcbs[h] = (h,r,c,b)
-                   in hrcbs
-                 let hrcbs =
-                   loop hrcbs = hrcbs for j < length X12 do
-                   let (h,c,b) = X12[j]
-                   let hrcbs[h] = (h,r,c,b)
-                   in hrcbs
-                 let D' = map (\(_,r,A) ->
-                                 map (\(_,c,B) ->
-                                        (idx_flatten nb (r,c),r,c,matmul A B)
-                                     ) X12
-                              ) X21 |> flatten
-                 let D = filter (\(_,r,c,_) -> r > i && c > i) blks
-                         |> map (\(h,r,c,b) -> (h,idx_flatten nb (r,c),r,c,b))
-                 let D'' = setops.join_by_key (.1) (.0) D D'                   -- ignore fillins
-                           |> map (\((h,_,r,c,b),(_,_,_,b')) -> (h,r,c,matsub b b'))
-                 let hrcbs =
-                   loop hrcbs = hrcbs for j < length D'' do
-                   let (h,r,c,b) = D''[j]
-                   let hrcbs[h] = (h,r,c,b)
-                   in hrcbs
-                 in hrcbs)
+      loop hrcbs for i < nb
+      do let blks = filter (\(_, r, c, _) -> r >= i && c >= i) (copy hrcbs)
+	 let (h, r, c, b) = blks[0] -- h is the idx into hrcbs identifying the current diagonal block
+         let b = lu.lu 1 b
+  	 let () = assert (r == i && c == i) ()
```

Comment:
This won't work. I'd put this `assert` around the RHS of the `A21` binding instead.